chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,347 @@
+// IMR: 0: The corresponding interrupt is disabled.
+//      1: The corresponding interrupt is enabled.
+// ISR: 0: An interrupt event has not occurred.
+//      1: An interrupt event has occurred.
+// ICR: Writing a one to this bit will clear the corresponding bit in ISR.
+// MODE:    0: The external interrupt is edge triggered.
+//          1: The external interrupt is level triggered.'
+// EDGE:    0: The external interrupt triggers on falling edge.
+//          1: The external interrupt triggers on rising edge.
+// LEVEL:   0: The external interrupt triggers on low level.
+//          1: The external interrupt triggers on high level.
+// FILTER:  0: The external interrupt is not filtered.
+//          1: The external interrupt is filtered.
+// ASYNC:   0: The external interrupt is synchronized to CLK_SYNC.
+//          1: The external interrupt is asynchronous.
+// EN: Writing a one to this bit will enable the corresponding external interrupt.
+// DIS: Writing a one to this bit will disable the corresponding external interrupt.
+// CTRL:    0: The corresponding external interrupt is disabled.
+//          1: The corresponding external interrupt is enabled.
+register_bitfields![
+    u32,
+    Interrupt [
+        /// Each bit represents the setup for each line, for sam4l, only bit 0-8 makes sense
+        INT OFFSET(0) NUMBITS(32) []
+    ],
+    /// Test is not being used right now
+    Test [
+        /// 0: This bit disables external interrupt test mode.
+        /// 1: This bit enables external interrupt test mode.
+        TESTEN OFFSET(31) NUMBITS(1) [],
+
+        /// Writing a zero to this bit will set the input value to INTn to zero, if test mode is enabled. 
+        /// Writing a one to this bit will set the input value to INTn to one, if test mode is enabled.
+        INT OFFSET(0) NUMBITS(31) []
+    ]
+];
+
+// Page 59 of SAM4L data sheet
+const EIC_BASE: StaticRef<EicRegisters> =
+    unsafe { StaticRef::new(0x400F1000 as *const EicRegisters) };
+
+impl PeripheralManagement<pm::Clock> for Eic<'a> {
+    type RegisterType = EicRegisters;
+
+    fn get_registers(&self) -> &EicRegisters {
+        &*EIC_BASE
+    }
+
+    fn get_clock(&self) -> &pm::Clock {
+        &Clock::PBD(PBDClock::EIC)
+    }
+
+    fn before_peripheral_access(&self, clock: &pm::Clock, _: &EicRegisters) {
+        clock.enable();
+    }
+
+    fn after_peripheral_access(&self, clock: &pm::Clock, registers: &EicRegisters) {
+        if registers.imr.get() == 0 && registers.ctrl.get() == 0 {
+            clock.disable();
+        }
+    }
+}
+
+type EicRegisterManager<'a, 'b> = PeripheralManager<'a, Eic<'b>, pm::Clock>;
+
+pub struct Eic<'a> {
+    callbacks: [OptionalCell<&'a dyn hil::eic::Client>; 9],
+}
+
+impl<'a> hil::eic::ExternalInterruptController for Eic<'a> {
+    type Line = Line;
+
+    fn line_enable(&self, line: &Self::Line, interrupt_mode: hil::eic::InterruptMode) {
+        let regs = &EicRegisterManager::new(&self).registers;
```

Comment:
```suggestion
        let regs = self.get_registers;
```

Reply:
I guess that it is not that costly to create `PeripheralManager::new` but it is already implemented and would be better just to call the method instead of creating the new value and just use it for the registers.

Reply:
Yea, this is better! I'm fixing these right now. Thanks!